src/utils/client.go

```diff
@@ -0,0 +1,264 @@
+	}
+	err = json.Unmarshal(content, &profile)
+	if err != nil {
+		return nil, err
+	}
+	return &profile, nil
+}
+
+func readTokenSalt(tokenSaltFile string) string {
+	var token string
+	f, err := os.Open(tokenSaltFile)
+	if os.IsNotExist(err) {
+		token, err = tokenSalt(tokenSaltFile)
+		if err != nil {
+			return ""
+		}
+		return token
+	}
+	defer f.Close()
+	content, _ := ioutil.ReadFile(tokenSaltFile)
+	token = string(content)
+
+	if strings.TrimSpace(token) == "" {
+		token, err = tokenSalt(tokenSaltFile)
+	}
+	return token
+}
+
+func tokenSalt(tokenSaltFile string) (string, error) {
+	token := randString(8)
+	err := ioutil.WriteFile(tokenSaltFile, []byte(token), 0644)
+	if err != nil {
+		return "", nil
+	}
+	return token, nil
+}
+
+func stringWithCharset(length int, charset string) string {
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[seededRand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+func randString(length int) string {
+	return stringWithCharset(length, charset)
+}
+
+func createHttpClient(caCertFile string, insecure bool) *http.Client {
+	// Get the SystemCertPool, continue with an empty pool on error
+	rootCAs, _ := x509.SystemCertPool()
+	if rootCAs == nil {
+		rootCAs = x509.NewCertPool()
+	}
+	// Read in the cert file
+	certs, err := ioutil.ReadFile(caCertFile)
+	if os.IsNotExist(err) {
+		Trace(fmt.Sprintf("The cert file is not found %v, ignore it", caCertFile))
+	} else {
+		// Append our cert to the system pool
+		if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
+			Trace("No certs appended, using system certs only")
+		}
+	}
+	// Trust the augmented cert pool in our client
+	config := &tls.Config{
+		InsecureSkipVerify: insecure,
+		RootCAs:            rootCAs,
+	}
+	tr := &http.Transport{TLSClientConfig: config}
+	return &http.Client{Transport: tr}
+}
+
+func (c *ConfigClient) basicAuth() string {
+	auth := c.username + ":" + c.password
+	return base64.StdEncoding.EncodeToString([]byte(auth))
+}
+
+func (c *ConfigClient) configHarborWithSetting(setting map[string]interface{}) error {
+	url := c.harborUrl + "/api/v2.0/configurations"
+	method := "PUT"
+	payload, err := json.Marshal(setting)
+	checkError(err)
+	req, err := http.NewRequest(method, url, bytes.NewReader(payload))
+	checkError(err)
+
+	req.Header.Add("Authorization", "Basic "+c.basicAuth())
+	req.Header.Add("Content-Type", "text/plain")
+	res, err := c.client.Do(req)
+	checkError(err)
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	checkError(err)
+	Trace(fmt.Sprintf("Response code %v\n", res.StatusCode))
+	if res.StatusCode != 200 {
+		return fmt.Errorf("failed to send request to %v, response code %v", url, res.StatusCode)
+	}
+	Trace("request content is " + string(body) + "\n")
+	return nil
+}
+
+func (c *ConfigClient) configOIDC(setting *UAAProfile, uaaUrl string, verifyCert bool) error {
+	s := map[string]interface{}{}
+	s["auth_mode"] = "oidc_auth"
+	s["oidc_endpoint"] = strings.ToLower(uaaUrl) + "/oauth/token"
```

Comment:
I think this is wrong.  oidc_endpoint should not carry the `/oauth/token` suffix.